README.md
# node-red-contrib-ezo

This Node-Red is designed to communicate with Atlas Scientific Ezo modules over I2C.

## Install

Either use the Manage Palette option in the Node-RED Editor menu, or run the following command in your Node-RED user directory - typically `~/.node-red`

    npm install --unsafe-perm node-red-contrib-ezo

## Usage

Designed to accept all standard commands from the Atlas Scientific documentation.

### Input

All commands can be sent as a single string in `msg.payload`, those which have a command and input value separated by a comma can be broken into `msg.command` and `msg.payload` properties. Commands can differ from one Ezo module to another, please refer to Atlas Scientific's website for documentation.  Below are a few examples of commands structured in an acceptable manner.  

Since v1.3.0 `msg.address` can be set dynamically on input.
Since v1.4.0 `msg.bus` can be set dynamically on input.

```javascript
msg = { payload: 'L,0' };  // turns led off

msg = { command: 'L', payload: 1 };  // turns led on

msg = { command: 'L', payload: false };  // turns led off

msg = { command: 'L', payload: '1' };  // turns led on

msg = { payload: 'Status' };  // returns status

msg = { command: 'R'};  // reads probe

msg = { command: 'R', address: 105, bus: 1 }  // reads probe at address 105 and bus 1
```

### Ouptut

Some commands such as 'Sleep', 'Factory', and 'I2C,n' do not provide a response so will not generate an output message.  Most commands will generate a response containing the original command and a value, however some Ezo modules can generate multiple values in which the `msg.payload` will contain an array.  If the output value can be parsed to a number it will be, otherwise it will be a string.  Since v1.1.1 extra properties in input pass through to output. 

```javascript
msg = {
    topic: String,
    command: String,
    payload: String / Number / Array,
}
```

Examples of commands and responses:

**Ezo Board:**  Temperature  
**Command:**  Status  
**Raw Response:**  1?STATUS,P,3.28  
  
**Node Output:**
```javascript
msg = {
    topic: 'tempurature',
    command: 'STATUS',
    payload: [
        'P',
        '3.28',
    ],
};
```
  
**--- OR ---**  
  
**Ezo Board:**  pH  
**Command:**  R   
**Raw Response:**  19.831
  
**Node Output:**
```javascript
msg = {
    topic: 'ph',
    command: 'R',
    payload: 9.831,
};
```

### Errors

Since v1.1.0 errors are catchable by the CATCH node.  If an error occurs, a message will not be sent.  Common errors include the following:

**'Invalid payload!'** - Error: Command must be under 32 characters.  
**'Syntax Error!'** - Error: Ezo board did not recognize the command.  
**'No data to send.'** - Warning: Command sent to the Ezo board was empty.  

## Important Note

This node uses the I2C-bus package from @fivdi. You can find his work on github: https://github.com/fivdi/i2c-bus  

CHANGELOG.md
v1.4.0 / December 26 2021
===================
  * Adds dynamic bus

v1.3.0 / October 27 2021
===================
  * Adds dynamic address to input (#5)
  * Fixes issue with input properties

v1.2.1 / September 18 2021
===================
  * Improved output data types
  * Fixes #2
  * Fixes #3

v1.2.0 / June 20 2021
===================
  
  * Adds customizable `msg.topic`

v1.1.2 / May 29 2021
===================
  
  * Improved node label text
  * Added `msg.topic` to output if no topic on input
  
v1.1.1 / May 27 2021
===================
  
  * Fixes #1

v1.1.0 (May 26 2021)
===================
  ### Added
  * Support for separate command/payload input properties.
  * Support for topic passthrough.
  * Supports error catching with catch node.

  ### Changed
  * Removed msg.status property in lue of better error handling / catch node.
  * Improved README.md
  
v1.0.2 / May 25 2021
===================
  
  * Fix NaN issue in payload

v1.0.1 / May 24 2021
===================
  
  * Fix error handling issue

v1.0.0 / May 24 2021
===================
  
  * Initial release

ezo.js
module.exports = function(RED) {
    var I2C = require("i2c-bus");

    function Ezo(config) {
        RED.nodes.createNode(this, config);
        this.ezoTopic = config.ezoTopic;
        //add default value for older nodes (<= 1.3.0)
        this.bus = config.bus || 1;
        if (config.customAddr) {
            this.address = parseInt(config.address);
        } else {
            switch(config.ezoBoard) {
                case 'ph':
                    this.address = 99;
                    break;
                case 'ec':
                    this.address = 100;
                    break;
                case 'orp':
                    this.address = 98;
                    break;
                case 'do':
                    this.address = 97;
                    break;
                case 'rtd':
                    this.address = 102;
                    break;
                case 'co2':
                    this.address = 105;
                    break;
                case 'o2':
                    this.address = 108;
                    break;
                case 'hum':
                    this.address = 111;
                    break;
                case 'prs':
                    this.address = 106;
                    break;
                case 'pmp':
                    this.address = 103;
                    break;
                case 'pmp-l':
                    this.address = 109;
                    break;
                case 'rgb':
                    this.address = 112;
                    break;
                case 'flo':
                    this.address = 104;
                    break;
            }
        }
        var node = this;
        var isRunning = false;
        var noRead = ['sleep', 'factory', 'i2c'];
        node.errorHandler = (port, err, msg) => {
            port.closeSync();
            node.error(err, msg);
            isRunning = false;
            return;
        }
        node.processRequest = (req) => {
            var newStr = '';
            var c = '';
            var p = '';
            if (req.hasOwnProperty('command')) { c = req.command.toString(); }
            if (req.hasOwnProperty('payload')) {
                p = req.payload.toString();
                if (p === 'true') p = '1';
                if (p === 'false') p = '0';
            }
            if (c !== '' && p !== '') {
                newStr = `${c},${p}`;
            } else if (c !== '') {
                newStr = c;
            } else if (p !== '') {             
                newStr = p;
            }
            return newStr;
        }
        node.processResponse = (res) => {
            var converted = { command: '', value: ''};
            var resString = res.toString('utf8', 1).replace(/\0/g, '');
            if (resString !== '') {
                if (resString.indexOf(',') !== -1) {
                    var splt = resString.split(',');
                    var stp = 0;
                    if (splt[0].indexOf('?') !== -1) {
                        converted.command = splt[0].replace('?', '');
                        stp = 1;
                    }
                    converted.value = [];
                    for (var i = stp; i < splt.length; i++) {
                        if (splt[i] === '') { 
                            converted.value.push( '' );
                        } else {
                            converted.value.push( isNaN(splt[i]) ? splt[i] : parseFloat(splt[i]) );
                        };
                    }
                } else {
                    converted.value = isNaN(resString) ? resString : parseFloat(resString);
                }
            } else {
                converted.command = '';
                converted.value = '';
            }
            return converted;
        };
        node.on("input", function(msg) {
            var pload = node.processRequest(msg);
            var addr = msg.address || node.address;
            var bus = msg.bus || node.bus;
            if (pload.length > 32) {
                node.error('Invalid payload!');
                return;
            }
            var buf = Buffer.from(pload);
            var port = I2C.openSync(bus);
            port.i2cWrite(addr, buf.length, buf, function(err) {
                if (err) {
                    node.errorHandler(port, err, msg);
                    return;
                };
                const strTest = new RegExp(noRead.join('|')).test(pload.toLowerCase());
                if (!strTest && !isRunning) {
                    isRunning = true;
                    var rBuf = Buffer.alloc(32);
                    const loop = setInterval(() => {
                        port.i2cRead(addr, rBuf.length, rBuf, function(err, size, res) {
                            if (err) {
                                node.errorHandler(port, err, msg);
                                return;
                            };
                            if (res[0] !== 254) {
                                clearInterval(loop);
                                if (res[0] === 2) {
                                    node.errorHandler(port, 'Syntax error!', msg);
                                    return;
                                }
                                port.closeSync();
                                if (res[0] === 255) {
                                    node.warn('No data to send.');
                                    isRunning = false;
                                    return;
                                }
                                var newRes = node.processResponse(res);
                                if (newRes.command === '') { newRes.command = (pload.indexOf(',') === -1) ? pload : pload.split(',')[0]; }
                                if (!msg.hasOwnProperty('topic') && node.ezoTopic !== undefined) { msg.topic = node.ezoTopic; }
                                msg.command = newRes.command;
                                msg.payload = newRes.value;
                                node.send(msg);
                                isRunning = false;
                            }
                        });
                    }, 300);
                }
            });
        });
    }
    RED.nodes.registerType("ezo", Ezo);
}